packages/wal/wal.go

```diff
@@ -0,0 +1,238 @@
+	if err := os.MkdirAll(w.dir, 0o777); err != nil {
+		return nil, fmt.Errorf("create dir: %w", err)
+	}
+	// read all segments in log
+	f, err := os.Open(w.dir)
+	if err != nil {
+		return nil, fmt.Errorf("could not open wal: %w", err)
+	}
+	var segments []*segment
+	files, _ := f.ReadDir(-1)
+	for _, file := range files {
+		w.metrics.segments.Inc()
+		index, _ := strconv.ParseUint(file.Name(), 10, 32)
+		segments = append(segments, &segment{index: uint32(index), dir: w.dir})
+	}
+	sort.SliceStable(segments, func(i, j int) bool {
+		return segments[i].index < segments[j].index
+	})
+	var lastIndex uint32
+	w.segments = segments
+	if len(segments) > 0 {
+		last := segments[len(segments)-1]
+		w.metrics.latestSegment.Set(float64(last.index))
+		lastIndex = last.index
+	}
+	return &chainWAL{w, chainID, lastIndex}, nil
+}
+
+func (w *chainWAL) Write(bytes []byte) error {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+
+	var index uint32 = 1
+	if len(w.segments) > 0 {
+		index = w.segments[len(w.segments)-1].index + 1
+	}
+	segment, err := w.createSegment(index)
+	if err != nil {
+		w.metrics.failedWrites.Inc()
+		return fmt.Errorf("Error writing log: %w", err)
+	}
+	n, err := segment.Write(bytes)
+	if err != nil || len(bytes) != n {
+		w.metrics.failedReads.Inc()
+		return fmt.Errorf("Error writing log: %w", err)
+	}
+	w.metrics.segments.Inc()
+	w.metrics.latestSegment.Set(float64(index))
+	return segment.Close()
+}
+
+func (w *chainWAL) createSegment(i uint32) (*segment, error) {
+	segName := segmentName(w.dir, i)
+	f, err := os.OpenFile(segName, os.O_CREATE|os.O_APPEND|os.O_WRONLY, 0o666)
+	if err != nil {
+		return nil, fmt.Errorf("could not create segment: %w", err)
+	}
+	s := &segment{index: i, segmentFile: f, dir: w.dir, name: segName}
+	w.segments = append(w.segments, s)
+	return s, nil
+}
+
+func segmentName(dir string, index uint32) string {
+	return filepath.Join(dir, fmt.Sprintf("%010d", index))
+}
+
+func (w *chainWAL) Contains(i uint32) bool {
+	for _, seg := range w.segments {
+		if seg.index == i {
+			return true
+		}
+	}
+	return false
+}
+
+func (w *chainWAL) Read(i uint32) ([]byte, error) {
+	segment := w.getSegment(i)
+	if segment == nil {
+		return nil, fmt.Errorf("block not found in wal")
+	}
+	if err := segment.load(); err != nil {
+		w.metrics.failedReads.Inc()
+		return nil, fmt.Errorf("Error opening backup file: %w", err)
+	}
+	stat, err := segment.Stat()
+	if err != nil {
+		w.metrics.failedReads.Inc()
+		return nil, fmt.Errorf("Error reading backup file: %w", err)
+	}
+	blockBytes := make([]byte, stat.Size())
+	bufr := bufio.NewReader(segment)
+	n, err := bufr.Read(blockBytes)
+	if err != nil || int64(n) != stat.Size() {
+		w.metrics.failedReads.Inc()
+		return nil, fmt.Errorf("Error reading backup file: %w", err)
+	}
+	return blockBytes, nil
+}
+
+func (w *chainWAL) getSegment(i uint32) *segment {
+	for _, seg := range w.segments {
+		if seg.index == i {
+			return seg
+		}
+	}
+	return nil
```

Comment:
Are segments sorted by their index? If so, you do not need to do full scan:
```suggestion
        if len(seg) == 0 {return nil}
        if (seg[0].index <= i) && (i <= seg[len(seg)-i].index) {
            return seg[i-seg[0].index]
        }
        return nil
```
If not, maybe you should use a map to sort it easily to access them quicker: `map[int]segment`, where `int` is segment index

Reply:
Wait, no. Probably my suggested code is not going to work. It assumes, all the blocks have consecutive indexes and such assumption is probably not possible. But I still think map would be better than simple array.